library-best-practices/JLBP-7.md

```diff
@@ -0,0 +1,35 @@
+[JLBP-7] Make breaking transitions easy
+---------------------------------------
+
+When a library makes an in-place breaking change (meaning that the Java package
+is not renamed - see [JLBP-6](JLBP-6.md) for more discussion on renaming rules),
+this can create a lot of work for consumers to adapt to the change. If a change
+is "atomic" (meaning that the old version of something like a method is removed
+at the same time the new version is added), all usages of the old method must be
+converted to the new method in a single change to each consumer's codebase. If
+the usage is wide enough and especially if it spans multiple packages, the cost
+might be too high to adopt at all.
+
+Consequently, libraries should instead make two-phase breaking changes to stable
+features:
+
+1. Mark old methods/classes as `@Deprecated` at the same time as adding new
+   methods/classes. This is a "stepping stone" release.
+2. Delete the deprecated methods/classes. This wouldn't necessarily be in the
+   next release, particularly if major consumers haven't updated yet.
```

Comment:
Given a 6-week release cycle, I'd strongly state that it _shouldn't_ be the next release. Fortunately, once you're already doing the change non-atomically, it's usually not too long to postpone the removal longer.

You do say this later on, but it could be nice to make clear from the beginning (assuming that doing so isn't too awkward).

Reply:
I don't fully grok why the 6-week release cycle would influence the decision. It seems like the adoption rate by consumers completely determines the course of action, regardless of the release cycle of the library making the breakage.

Reply:
A 6-week cycle gives consumer 6 weeks to adopt the new thing. If everyone does release every 6 weeks, that's OK, but that seems optimistic.

Again, though, it seems like you're already saying what I'm advocating for later on, so maybe I'm misunderstanding what you're saying here?

Reply:
I have clarified the conditions for deleting the deprecated stuff. (PTAL)

The reason I want to leave the release cycle out of that is that the release cycle will be suggested by a separate best practice, and I don't want the bast practices unnecessarily coupled. Anyway, the condition of adoption is the key condition that needs to be met, and talking about release cycles is just a proxy anyway.